---
title: "scratch"
author: "Keon Arbabi"
date: "03/03/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# load libraries 
```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(plyr)
  library(broom)
  library(data.table)
  library(GEOquery)
  library(MDSeq)
  library(edgeR)
  library(limma)
  library(HelpersMG)
  library(Seurat)
  library(glmGamPoi)
  library(matrixStats)
  library(gsEasy)
  library(hypeR)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(ggpubr)
  library(grid)
  library(gridExtra)
  library(cowplot)
  library(ggbeeswarm)
  library(qusage)
  library(gplots)
  library(colorspace)
  })
```

# functions
```{r}
convertMouseGeneList = function(x){
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
  genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = x , mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
  mousex = unique(genesV2[, 2])
  # Print the first 6 genes found to the screen
  print(head(mousex))
  return(genesV2)
}

# https://github.com/satijalab/sctransform/blob/master/R/utils.R
row_gmean = function(x, eps = 1) {
  if (inherits(x = x, what = 'matrix')) {
    return(exp(rowMeans(log(x + eps))) - eps)
  }
  if (inherits(x = x, what = 'dgCMatrix')) {
    ret = row_gmean_dgcmatrix(matrix = x, eps = eps)
    names(ret) = rownames(x)
    return(ret)
  }
  stop('matrix x needs to be of class matrix or dgCMatrix')
}

reverselog_trans = function(base=exp(1)) {
    trans = function(x) -log(x, base)
    inv = function(x) base^(-x)
    scales::trans_new(paste0("reverselog-", format(base)), trans, inv, 
              scales::log_breaks(base=base), 
              domain=c(1e-100, Inf))
}

```

# download data
```{r}
if(F){
  wget(c("http://pklab.med.harvard.edu/viktor/publications/Epilepsy19/count_matrices.rds",
         "http://pklab.med.harvard.edu/viktor/publications/Epilepsy19/con_filt_cells.rds",
         "http://pklab.med.harvard.edu/viktor/publications/Epilepsy19/con_filt_samples.rds"))
}
```

# prepare data 
```{r}
# read sample metadata
samples_metadata = read.csv(url("https://raw.githubusercontent.com/khodosevichlab/Epilepsy19/master/metadata/sample_info.csv")) %>%
  mutate(Condition = case_when(Alias %in% c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10") ~ "Control",
                               Alias %in% c("E1","E2","E3","E4","E5","E6","E7","E8","E9") ~ "Epilepsy"))
# read cell annotation metadata 
annot_metadata = read.csv(url("https://raw.githubusercontent.com/khodosevichlab/Epilepsy19/master/metadata/annotation.csv"))
rownames(annot_metadata) = annot_metadata$cell

# read counts
counts_lst = readRDS(file = "./input/count_matrices.rds")
counts_lst = counts_lst[names(counts_lst) != "NeuN"]
# get number of cells from each subject and add to metadata
samps = lapply(counts_lst, function(x){
  return(data.frame(cells = ncol(x)))
})
samps = do.call(rbind, samps) %>% rownames_to_column(var = "Name")
samples_metadata = samples_metadata[match(samps$Name, samples_metadata$Name),]
samples_metadata$N_cells = samps$cells
rownames(samples_metadata) = NULL
samps = c(rep(samps[,1], times = samps[,2]))

# convert to sparse matrix
counts_matrix = do.call(cbind, counts_lst)
dim(counts_matrix)
cells = colnames(counts_matrix)
# rename columns using unique sample id
colnames(counts_matrix) = paste(samps, cells, sep = "_")

# combine metadata
samples_metadata = samples_metadata[rep(1:nrow(samples_metadata), samples_metadata$N_cells),]
samples_metadata$cell = colnames(counts_matrix)
metadata = left_join(annot_metadata, samples_metadata, by = "cell")
rownames(metadata) = metadata$cell

# filter to match annotation metadata 
counts_matrix = counts_matrix[,colnames(counts_matrix) %in% metadata$cell]
dim(counts_matrix)
metadata = metadata %>% filter(cell %in% colnames(counts_matrix))
write.csv(metadata, file = "./output/metadata.csv")

# create Seurat object
seu_obj = CreateSeuratObject(counts = counts_matrix, meta.data = metadata)
head(x = rownames(x = seu_obj))
head(x = colnames(x = seu_obj))
rm(counts_lst, counts_matrix)
gc()

# # normalize
# # https://doi.org/10.1186/s13059-019-1874-1
# seu_obj_norm = SCTransform(seu_obj, method = "glmGamPoi", verbose = TRUE, return.only.var.genes = FALSE)
# # save Seurat object
# saveRDS(seu_obj_norm, file = "./output/seu_obj_norm.rds")

```

# calculate intra-individual gene variance 
```{r}
# load normalized Seurat object 
seu_obj_norm = readRDS(file = "./output/seu_obj_norm.rds")
metadata = read.csv(file = "./output/metadata.csv", row.names = "X")
table(metadata$Name, metadata$l2)
# remove subject with low cell counts 
seu_obj_norm = subset(seu_obj_norm, subset = Name == "HB51", invert = TRUE)
metadata = metadata %>% filter(!Name == "HB51")
# initialize
res = NULL

# get model attributes
sct_model_df = seu_obj_norm@assays$SCT@SCTModel.list$model1@feature.attributes
#write.csv(sct_model_df, file = "./output/sct_model_df.csv")

# split cells by subclass and subject
# this amounts to calculating intra-individual variability within cells types 
for(CT in unique(metadata$l2)){
  for(INDV in unique(metadata$Name)){
    # subset
    seu_obj_tmp = subset(seu_obj_norm, subset = l2 == CT & Name == INDV)
    # get normalized expression matrix (Pearson residuals)
    sct_matrix = seu_obj_tmp@assays$SCT@scale.data %>% as.matrix() 
    dim(sct_matrix)
    # get normalized counts matrix
    norm_counts_matrix = seu_obj_tmp@assays$SCT@counts %>% as.matrix()
    dim(norm_counts_matrix)
    # get original counts matrix
    counts_matrix = seu_obj_tmp@assays$RNA@counts %>% as.matrix()
    counts_matrix = counts_matrix[match(rownames(norm_counts_matrix),rownames(counts_matrix)), 
                                  match(colnames(norm_counts_matrix),colnames(counts_matrix))]
    dim(counts_matrix)
    # get gene-wise coefficient of variation (Stdev/Mean*100)
    tmp = data.frame(Name = INDV,
                     Cell_type = CT,
                     Gene = rownames(sct_matrix),
                     N_cells = ncol(sct_matrix),
                     N_NonZero = rowSums(norm_counts_matrix != 0),
                     Pct_NonZero = rowSums(norm_counts_matrix != 0)/ncol(norm_counts_matrix),
                     Counts = rowSums(counts_matrix),
                     Counts_norm = rowSums(norm_counts_matrix),
                     Mean = rowMeans(sct_matrix),
                     Var = rowVars(sct_matrix),
                     Stdev = rowSds(sct_matrix)) %>%
      mutate(CV = Stdev/(Mean*100))
    # save
    res = rbind(res, tmp)
    # # visualize gene mean-variance relationship
    # tmp %>%
    #   as.data.frame() %>%
    #   ggplot(., aes(x = log10(Mean), y = Var)) +
    #   geom_point(alpha = 0.5, fill = "grey", colour = "black") +
    #   geom_density_2d(size = 0.3) +
    #     labs(x = "Log10(Mean expression)",
    #          y = "Residual variance") +
    #     ggtitle("Gene Mean-Variance Relationship")
    # ggsave(filename = paste0("./figures/qc/sct_",INDV,"_",CT,".jpeg"), width = 6, height = 5.5)
  }
}
res = left_join(res, distinct(metadata[,c(6:10)]), by = "Name") 
# save results 
#write.csv(res, file = "./output/res.csv")

```

# comparison 
```{r}
# load 
res = fread(file = "./output/Pfisterer/res.csv", data.table = FALSE) %>% column_to_rownames(var = "V1")
res_filt = res %>% filter(N_NonZero > 0)

# distributions
p1 = ggplot(res_filt, aes(x = Mean)) + # mean
  geom_histogram(bins = 100) + geom_vline(xintercept = median(res_filt$Mean), color = "red") + scale_x_continuous(limits = c(-1, 1)) + theme_classic()
p2 = ggplot(res_filt, aes(x = Var)) + # variance
  geom_histogram(bins = 100) + geom_vline(xintercept = median(res_filt$Var), color = "red") + scale_x_continuous(limits = c(-0.5, quantile(res_filt$Var, 0.99))) + theme_classic()
p3 = ggplot(res_filt, aes(x = Stdev)) + # standard deviation
  geom_histogram(bins = 100) + geom_vline(xintercept = median(res_filt$Stdev), color = "red") + scale_x_continuous(limits = c(-0.5, quantile(res_filt$Stdev, 0.99))) + theme_classic()
ggplot(res_filt, aes(x = CV)) + # coefficient of variation
  geom_histogram(bins = 150) + geom_vline(xintercept = median(res_filt$CV), color = "red") + scale_x_continuous(limits = c(-1, 1)) + theme_classic()
ggarrange(p1, p2, p3, nrow = 1)
ggsave(filename = "./figures/distributions.png", width = 9, height = 4)

# by cell
ggplot(res_filt, aes(x = Mean)) +
  geom_histogram(bins = 100) + geom_vline(xintercept = median(res_filt$Mean), color = "red") + scale_x_continuous(limits = c(-1, 1)) + theme_classic() + facet_wrap(~ Cell_type)
ggplot(res_filt, aes(x = Stdev)) + 
  geom_histogram(bins = 100) + geom_vline(xintercept = median(res_filt$Stdev), color = "red") + scale_x_continuous(limits = c(-0.5, quantile(res_filt$Stdev, 0.99))) + theme_classic() + facet_wrap(~ Cell_type)

# get genes that are correlates of electrophysiology
# https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005814
ephys_genes = read.csv(file = "./input/Tripathy_2017_ephys_genes.csv")[,-1] %>% dplyr::rename(MGI.symbol = GeneSymbol)
# convert from mouse genes 
hgnc = convertMouseGeneList(ephys_genes$MGI.symbol)
ephys_genes = left_join(ephys_genes, hgnc, by = "MGI.symbol") %>% 
  filter(EphysProp %in% c("Rheo","Vrest"), HGNC.symbol %in% intersect(HGNC.symbol, res_filt$Gene)) %>% pull(HGNC.symbol) %>% unique()

# get genes coding proteins with detected expression in human brain
# https://www.proteinatlas.org/search/NOT+tissue_category_rna%3Abrain%3BNot+detected
brain_genes = read_tsv(file = "./input/tissue_category_rna_brain_Tissue_detected.tsv", show_col_types = F) %>% 
  filter(Gene %in% intersect(Gene, res_filt$Gene)) %>% pull(Gene) %>% unique()
random_genes = sample(brain_genes, length(ephys_genes))

plot_df = res_filt %>%
  #filter(Gene %in% c(ephys_genes, random_genes)) %>%
  mutate(Gene_group = case_when(Gene %in% ephys_genes ~ "Vrest & rheobase associated genes (n = 277)", 
                                Gene %in% random_genes ~ "Random brain-expressed genes (n = 277)", TRUE ~ "All genes (n = 28,479)")) %>% 
  group_by(Condition, Cell_type, Gene) %>%
  mutate(Mean_Stdev = mean(Stdev)) %>%
  distinct(Condition, Cell_type, Gene, .keep_all = TRUE) 

# compare expression sd 
ggplot(plot_df, aes(x = Condition, y = Mean_Stdev)) +
  geom_violin(aes(fill = Condition), alpha = 1) +
  geom_boxplot(fill = "white", outlier.shape= NA, width = 0.2) +
  scale_y_continuous(trans = "log10", limits = c(quantile(plot_df$Mean_Stdev, 0.01), quantile(plot_df$Mean_Stdev, 0.99))) +
  scale_fill_manual(breaks = c("Control","Epilepsy"), values = c("darkgreen","darkorange")) +
  labs(x = "condition", y = "standard deviation of expression") +
  coord_flip() +
  facet_wrap(~ Gene_group, ncol = 1) +
  theme_classic() +
  theme(legend.position = "none") +
  stat_compare_means(method = "wilcox.test", aes(group = Condition), label.y = 0.2, label.x = 1.5,  size = 3)
ggsave(filename = paste0("./figures/ephy_genes_sd_violin.jpeg"), width = 9, height = 6)

# compare expression sd by cell type
plot_df = plot_df %>% filter(!Gene_group == "All genes (n = 28,479)")
ggplot(plot_df, aes(x = Condition, y = Mean_Stdev)) + 
  geom_boxplot(aes(fill = Condition), outlier.shape = NA) +
  geom_quasirandom(na.rm = T, shape = 16, alpha = 0.1, size = 0.6) + 
  scale_y_continuous(trans = "log10", limits = c(quantile(plot_df$Mean_Stdev, 0.01), quantile(plot_df$Mean_Stdev, 0.99))) +
  scale_fill_manual(breaks = c("Control","Epilepsy"), values = c("darkgreen","darkorange")) +
  labs(x = "condition", y = "standard deviation of expression") +
  coord_flip() +
  facet_grid(Cell_type ~ Gene_group, scales = "free") +
  theme_classic() +
  theme(legend.position = "none") +
  stat_compare_means(ref.group = "Control", label = "p.signif", method = "wilcox.test") 
ggsave(filename = paste0("./figures/ephy_genes_sd_violin_ct.jpeg"), width = 9, height = 6.5)


# plot_df = res %>%
#   filter(Cell_type %in% c("L5_6_Fezf2","Pvalb")) %>%
#   filter(Pct_NonZero > 0.1, 
#          Gene %in% ephys_genes) 
# p2 = ggplot(plot_df, aes(x = Condition, y = Stdev, fill = Condition)) + 
#   geom_violin() + 
#   geom_boxplot(fill = "white", outlier.shape = NA, width = 0.4) +
#   scale_y_continuous(trans = "log10", limits = c(quantile(plot_df$Stdev, 0.003), quantile(plot_df$Stdev, 0.999))) +
#   scale_fill_manual(breaks = c("Control","Epilepsy"), values = c("darkgreen","darkorange")) +
#   labs(x = "Condition", y = "gene-wise standard deviation of expression") +
#   #coord_flip() +
#   facet_grid(~Cell_type, scales = "free_y") +
#   theme_classic() +
#   theme(legend.position = "none") +
#   stat_compare_means(method = "wilcox.test", label = "p.signif") 


# get patholophysiologically related GO terms and genes 
# sample to same set size
go_ecm_genes = read.table(file = "./input/GO_0030198.txt", fill = T)[2] %>% unlist() %>% sample(.,length(ephys_genes)) # extracellular matrix organization
go_immune_genes = read.table(file = "./input/GO_0002253.txt", fill = T)[2] %>% unlist() %>% sample(.,length(ephys_genes)) # activation of immune response
go_mapk_genes = read.table(file = "./input/GO_0043408.txt", fill = T)[2] %>% unlist() %>% sample(.,length(ephys_genes)) # regulation of MAPK cascade
go_axondev_genes = read.table(file = "./input/GO_0061564.txt", fill = T)[2] %>% unlist() %>% sample(.,length(ephys_genes)) # axon development
go_ap_genes = read.table(file = "./input/GO_0001508.txt", fill = T)[2] %>% unlist() # action potential

# get random GO terms and genes 
go_genes = read_rds(file = "./input/go_genes.rds")
# sample to roughly the same set size
keep = lengths(go_genes) > 250 & lengths(go_genes) < 300
go_genes = go_genes[keep]
go_rand_genes1 = sample(go_genes,1) %>% unlist()
go_rand_genes2 = sample(go_genes,1) %>% unlist()
go_rand_genes3 = sample(go_genes,1) %>% unlist()

gene_set_lst = list(ephys_genes, go_ecm_genes, go_immune_genes, go_mapk_genes, go_axondev_genes, go_ap_genes, go_rand_genes1, go_rand_genes2, go_rand_genes3)
names(gene_set_lst) = c("ephys_genes", "go_ecm_genes", "go_immune_genes", "go_mapk_genes", "go_axondev_genes", "go_ap_genes", "go_rand_genes1", "go_rand_genes2", "go_rand_genes3")

## Simulation 
# initialize 
res_sim = NULL
sim_lst = NULL
plot_lst = NULL
# loop through each condition and cell type
for(cx in unique(res_filt$Condition)){
  for(ct in unique(res_filt$Cell_type)){
    sim = NULL
    tmp = res_filt %>% filter(Condition == cx, Cell_type == ct) 
    for(i in 1:10000){
      # simulate distribution of average standard deviations for brain expressed genes
      null_mean = tmp %>% filter(Gene %in% sample(brain_genes, length(ephys_genes))) # random sampling 
      null_mean = mean(null_mean$Stdev)
      sim = c(sim, null_mean)
      print(paste(cx,ct,i))
    }
    # get average standard deviation for each gene set
    h_means = sapply(gene_set_lst, function(x){
      tmp %>% filter(Gene %in% x) %>% pull(Stdev) %>% mean
    })
    # empirical distribution function for calculating p-value (probability)
    # of the average standard deviation of expression for ephys genes in the simulated distribution
    d_fun = ecdf(sim)
    p_vals = d_fun(h_means)
    #plot(d_fun,  main = paste("empirical distribution function",10000,"iterations"))
    
    # save results
    tmp = data.table(Condition = cx, Cell_type = ct, Null_mean = null_mean, H_mean = h_means, P_val = p_vals, Shapiro = shapiro.test(sample(sim,5000))['p.value'], Gene_set = names(gene_set_lst))
    res_sim = rbind(res_sim, tmp)
    sim_lst[[paste(cx,ct)]] = sim
    # save plot
    plot_lst[[paste(cx,ct)]] = data.frame(val = sim) %>%
      ggplot(., aes(val)) +
      geom_histogram(aes(y = ..density..), bins = 100, fill = "grey") +
      stat_function(fun = dnorm, args = list(mean = mean(sim), sd = sd(sim))) +
      geom_vline(xintercept = mean(h_means[1]), color = "red") +
      labs(x = "St. Dev", y = "Count", title = paste(cx,ct,"\np-value:",p_vals[1])) +
      theme_classic()
  }
}
#write.csv(res_sim, file = "./output/res_sim.csv")
#png(filename = "./figures/simulations.png", width = 15, height = 9,  units = "in", res = 600)
grid.arrange(grobs = plot_lst, ncol = 4, nrow = ceiling(length(plot_lst)/4))
#dev.off()

res_sim = read.csv(file = "./output/res_sim.csv")

plot_df = res_sim %>%
  filter(!Gene_set %in% c("go_rand_genes1","go_rand_genes2","go_rand_genes3")) %>%
  dplyr::select(Condition, Cell_type, Gene_set, P_val) %>%
  unite(Group, Condition:Cell_type, remove = TRUE) %>%
  pivot_wider(id_cols = Group, names_from = Gene_set, values_from = P_val) %>%
  column_to_rownames(var = "Group") %>%
  as.matrix()
plot_df = plot_df[,order(colMeans(plot_df), decreasing = F)]

col = choose_palette()
col = viridis::viridis_pal()(20)
col = rev(col)

heatmap.2(plot_df, 
          scale = "none",
          Rowv = FALSE,
          Colv = FALSE,
          dendrogram = "none",
          col = col, trace = "none",
          cexCol = 0.9,
          cexRow = 0.9,
          margins = c(8,10),
          density.info = "histogram",
          key = TRUE, keysize = 1.2, key.title = "", key.xlab = "P-value", key.ylab = "")


```

# differential variability testing
```{r}
# https://academic.oup.com/nar/article/45/13/e127/3848359
res_dv = NULL
for(ct in unique(res$Cell_type)){
  # get pseudobulk counts
  print(ct)
  print(unique(res$Cell_type))
  dat = res %>%
    filter(Cell_type == ct) %>%
    dplyr::select(Alias, Gene, Counts) %>%
    pivot_wider(names_from = Alias, values_from = Counts) %>%
    column_to_rownames(var = "Gene")
  group = substring(names(dat), 1, 1)
  group = factor(group, levels = c("C","E"))
  
  # filter genes
  # data were already pre-filtered by Pfisterer et al. 
  dat_filtered = filter.counts(dat, mean.cpm.cutoff = 0.1)
  
  # normalize data
  # not sure if this is necessary as we're using scTransform normalized counts,
  # which accounts for differences in cell-to-cell sequencing depth 
  dat_normalized = normalize.counts(dat_filtered, method = "TMM")
  # contrasts 
  groups = get.model.matrix(group)
  # remove outliers
  dat_checked = remove.outlier(dat_normalized, contrast = groups, min.sample.size = 7, mc.cores = 10)
  counts = dat_checked$count[dat_checked$outliers$status==0,] 
  dim(counts)
  
  # differential mean and dispersion analysis 
  fit = MDSeq(counts, contrast = groups, mc.cores = 10)
  tmp = extract.ZIMD(fit, compare = list(A="E",B="C"), log2FC.threshold = 0, p.adj = "BH") %>%
    rownames_to_column(var = "Gene") %>%
    mutate(Cell_type = ct)
  res_dv = rbind(res_dv, tmp)
}
#write.csv(res_dv, file = "./output/res_dv.csv")

```

# visualize DV genes 
```{r}
#load 
res_dv = read.csv(file = "./output/res_dv.csv")[-1]

# p-value distributions 
par(mfrow=c(2,4))
dlply(res_dv, .(Cell_type), .fun = function(x){
    hist(x$Pvalue.dispersion, xlab = "P-value", main = paste(as.character(unique(x$Cell_type)),"median:",round(median(x$Pvalue.dispersion, na.rm = T),2)))
})
# get ephys gene enrichment scores from hypergeometric and permutation tests 
res_gse = NULL
res_hyp = NULL
for(ct in unique(res_dv$Cell_type)){
  tmp = res_dv %>%
    filter(Cell_type == ct) %>%
    mutate(Rank = sign(EvsC.dispersion.log2FC.0)*-log10(Pvalue.dispersion)) %>%
    arrange(Rank) 
  s = which(tmp$Gene %in% ephys_genes, arr.ind = TRUE)
  tmp = data.frame(Enrichment_score = gset(S = s, N = nrow(tmp), raw_score = TRUE),
                   P_val = gset(S = s, N = nrow(tmp), raw_score = FALSE), 
                   Cell_type = ct)
  res_gse = rbind(res_gse, tmp)
  
  tmp = res_dv %>% filter(Cell_type == ct)
  signature = tmp %>% filter(EvsC.dispersion.log2FC.0 < 0, FDR.dispersion < 0.01) %>% pull(Gene)
  geneset = intersect(tmp$Gene, ephys_genes) %>% list
  names(geneset) = "ephy_genes"
  
  tmp = hypeR(signature, geneset, test = "hypergeometric", plotting = T)
  tmp = tmp$data
  tmp$Cell_type = ct
  res_hyp = rbind(res_hyp, tmp)
}
# enrichment plot
ggplot(res_gse, aes(x = reorder(Cell_type, Enrichment_score), y = Enrichment_score, color = P_val)) +
  geom_point(shape =  16, alpha = 0.9, size = 9) +
  labs(y = "Enrichment score", x = "Cell type", title = "", color = "P-value") +
  coord_flip() +
  scale_x_discrete(labels = function(x) str_wrap(x, width = 45)) +
  scale_y_continuous(trans = "log10") +
  scale_color_continuous(low = "darkorange", high = "darkgreen", guide = guide_colorbar(reverse = TRUE)) +
  #geom_hline(yintercept = 0.05, linetype = "dotted") +
  guides(size = "none") + 
  theme_classic() 
ggsave(filename = "./figures/enrichment_plot.png", width = 5, height = 6)

# volcano plots 
plot_df = res_dv %>%
  mutate(Select_gene = case_when(Gene %in% ephys_genes ~ "TRUE", TRUE ~ "FALSE")) %>%
  filter(!-log10(Pvalue.dispersion) > 50)
ggplot(plot_df) +
  geom_point(data = plot_df %>% filter(Select_gene == "FALSE"), aes(x = EvsC.dispersion.log2FC.0, y = -log10(Pvalue.dispersion)), shape = 16, color = "grey", alpha = 0.4) +
  geom_point(data = plot_df %>% filter(Select_gene == "TRUE"), aes(x = EvsC.dispersion.log2FC.0, y = -log10(Pvalue.dispersion)), shape = 16, color = "darkorange", alpha = 0.7) +
  labs(x = "dispersion log2 fold-change", y = "dispersion p-value (-log10)") +
  facet_wrap(~ Cell_type, scales = "free") +
  theme_classic()
ggsave(filename = "./figures/volcano_plots.png", width = 9, height = 9)

```


```{r}
# get differentially expressed (DE) genes between case and controls in L4 Rorb neurons 
de_genes = res_dv %>% filter(Cell_type == "L4_Rorb", FDR.mean < 0.10) %>% pull(Gene)
# plot heatmap of DE genes 
dat = res %>%
  filter(Cell_type == "L4_Rorb", Gene %in% de_genes) %>%
  dplyr::select(Alias, Gene, Counts_norm) %>%
  pivot_wider(names_from = Alias, values_from = Counts_Norm) %>%
  column_to_rownames(var = "Gene") %>% as.matrix()
#col = choose_palette()
#png(filename = "./figures/de_heatmap.png", width = 8.5, height = 7, units = "in", res = 400)
heatmap.2(dat, 
          scale = "row",na.rm = T,
          Rowv = T, Colv = T, dendrogram = "both",
          col = col, trace = "none",
          cexCol = 0.9, cexRow = 0.9, srtCol = 45,
          margins = c(8,10),
          density.info = "histogram",
          key = TRUE, keysize = 1.2, key.title = "", key.xlab = "Residual Mean\nExpression", key.ylab = "")
#dev.off()

res_de = dat %>% 
  as.data.frame() %>%
  mutate(E_means = rowMeans(dplyr::select(., c("E1","E2","E7","E3","E8","E9")), na.rm = TRUE)) %>%
  mutate(C_means = rowMeans(dplyr::select(., starts_with("C")), na.rm = TRUE)) %>%
  dplyr::select(E_means, C_means) %>%
  rownames_to_column(var = "Gene")

# load human patch-seq data
# load summarized ephys features 
# ephys_b = read.csv(url('https://raw.githubusercontent.com/AllenInstitute/patchseq_human_L23/master/data/human_mouse_ephys_all_0127.csv'))
ephys_b = read.csv(file = "./input/human_ephys_features_berg.csv")
ephys_b = ephys_b %>% mutate(input_resistance = log10(input_resistance), 
                             rheobase_i = log10(rheobase_i), 
                             width = log10(width * 1000),
                             apamp = peak_v - threshold_v, 
                             ahpamp = threshold_v - fast_trough_v)
# load metadata
metadata_b = fread(file = "./input/20200625_patchseq_metadata_human.csv", data.table = FALSE) %>% 
  dplyr::rename(sample_id = transcriptomics_sample_id, cluster_label = corresponding_AIT2.3.1_alias) %>%
  filter(cluster_label %in% c("Exc L3-5 RORB COL22A1","Exc L3-4 RORB CARM1P1")) %>%
  filter(!sample_id %in% c("PS061517_E1-50_S20","PS061517_E1-50_S36","SM-GE65Y_S123_E1-50","SM-GE67G_S107_E1-50","SM-GE5PA_S137_E1-50")) %>%
  dplyr::select(c(ephys_session_id, sample_id, donor_id, cell_soma_normalized_depth, cluster_label, depth_from_pia_um)) %>%
  left_join(., ephys_b, by = "ephys_session_id")
# load expression 
counts_matrix_b = fread(file = "./input/20200512_Human_PatchSeq_Release_count.csv", data.table = FALSE) %>%
  dplyr::rename(gene = 1) %>%
  dplyr::distinct(gene, .keep_all= TRUE) %>%
  column_to_rownames(var = "gene") %>%
  filter_all(any_vars(. != 0))
metadata_b = metadata_b %>% filter(sample_id %in% colnames(counts_matrix_b))

de_genes = intersect(de_genes, rownames(counts_matrix_b))
counts_matrix_b = counts_matrix_b[de_genes, metadata_b$sample_id]
lcpm_matrix_b = edgeR::cpm(counts_matrix_b, log = TRUE, prior.count = 2)
res_de = res_de %>% filter(Gene %in% de_genes)

# get correlation with epilepsy DE genes for each patch-seq cell
cor_scores = data.frame(sample_id = metadata_b$sample_id, cor = apply(lcpm_matrix_b, MARGIN = 2, FUN = function(x) cor(x, res_de$E_means)))
metadata_b = metadata_b %>% left_join(., cor_scores, by = "sample_id") 



model_features_list = c('avg_rate','input_resistance','rheobase_i','width','apamp','ahpamp','threshold_v','trough_v','v_baseline','sag','adapt')
lm_df= NULL
for(f in model_features_list){
  formula = paste0('scale(', f, ') ~ scale(cor) + scale(depth_from_pia_um)')
  print(formula)
  tmp = tidy(lm(formula, data = metadata_b)) %>%
    mutate(term = recode(term, `scale(cor)` = "Epilepsy_cor", `scale(depth_from_pia_um)` = "Abs. depth")) %>%
    mutate(ephys_feature = f)
  lm_df = rbind(lm_df, tmp)
}
lm_df$ephys_feature = factor(lm_df$ephys_feature, 
                             levels = c("avg_rate","input_resistance","rheobase_i","v_baseline","sag","threshold_v","trough_v","apamp","ahpamp","width","fi_fit_slope","adapt"),
                             labels = c("Average rate","Input resistance","Rheobase","RMP","Sag ratio","AP thresh","AP trough","AP amp","AHP amp","AP width","FI slope","Adaptation"))
remove_ephys_feature = c("")

lm_df %>% filter(term == 'Epilepsy_cor', !ephys_feature %in% remove_ephys_props) %>% 
  ggplot(aes(x = reorder(ephys_feature, estimate), y = estimate)) + 
  geom_bar(stat = "identity", position = "dodge2", fill = "darkgreen") +
  geom_errorbar(aes(x = ephys_feature, y = estimate, ymin = estimate - std.error, ymax = estimate + std.error), position = "dodge2") +
  geom_hline(yintercept = 0) + 
  labs(x = "", y = "Std. Beta (epilepsy transcriptomic signature)") +
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
ggsave(filename = "./figures/lm_plot.png", width = 4, height = 6)

p1 = metadata_b %>% 
  ggplot(aes(x = width, y = cor)) +
  geom_point(shape = 16) +
  geom_smooth(method = "lm", se = F, color = "darkorange") +
  stat_cor(method = "pearson", size = 4) +
  labs(x = "AP width", y = "correlation with epilepsy transcriptional signature") +
  theme_classic()

p2 = metadata_b %>% 
  ggplot(aes(x = avg_rate, y = cor)) +
  geom_point(shape = 16) +
  geom_smooth(method = "lm", se = F, color = "darkorange") +
  stat_cor(method = "pearson", size = 4) +
  labs(x = "Averate rate", y = "correlation with epilepsy transcriptional signature") +
  theme_classic()

p3 = metadata_b %>% 
  ggplot(aes(x = sag, y = cor)) +
  geom_point(shape = 16) +
  geom_smooth(method = "lm", se = F, color = "darkorange") +
  stat_cor(method = "pearson", size = 4) +
  labs(x = "Sag ratio", y = "correlation with epilepsy transcriptional signature") +
  theme_classic()

ggarrange(p1,p2,p3, nrow = 1)
ggsave(filename = "./figures/cor_plots.png", width = 10, height = 6)


```











